### PR TITLE
[WIP] Cars, and clown cars and a ton of memes

### PR DESCRIPTION
**Pull request by ghost**
:cl: Floyd (and sprites by KMC)
rscadd: Adds cars
rscadd: Adds the tourist car that cargo can buy. You can use it to drive around in style.
rscadd: Adds the clown car, a 12TC meme vehicle for clown traitor only. It can be used to ram and kidnap people.
/:cl:

Adds obj/vehicle/car which is a more advanced version of normal vehicles that have action buttons and have functions like honking, you can easily add new actions so you could make stuff like car weapons.

as a proof of concept I also added clown cars which will be a clown antag item that function similar to Goons version. you can ram people and stuff em in your car till you crash or ya car is destroyed.

https://streamable.com/mdcvy

You can also emag cars to make it be able to ram, adds damage to ramming and lower the honk spam timer.

### CODE REVIEW

File: code/modules/vehicles/car/car.dm [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	if(auto_door_open && istype(M, /obj/machinery/door))
+		M.Bumped(driver)
+	if(ramming)
+		if(ishuman(M))
+			var/mob/living/carbon/human/H = M
+			src.visible_message("<span class='danger'>[src] rams into [H] and knocks them down!</span>")
+			H.Weaken(3)
+			if(ramming_sounds.len)
+				playsound(loc, pick(ramming_sounds), 75)
+		else if(!istype(M, /obj/machinery/door) && world.time - last_crash_time > 20 && istype(M, /obj) || istype(M, /turf/closed))
```
Comment (AuroranAI) [inline code/modules/vehicles/car/car.dm]: will always return true if it's a wall

Comment (ghost) [reply code/modules/vehicles/car/car.dm]: fixed in latest

**Comment by Kmc2000** [comment]
I came as soon as I saw this.

Would merge / :100:

File: code/modules/vehicles/car/car_actions.dm [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+/datum/action/innate/car/kidnap_count
+	name = "Kidnap Count"
+	button_icon_state = "car_count0"
+
+/datum/action/innate/car/kidnap_count/proc/update_counter()
+	var/count = car.loaded_humans.len
+	switch(count)
+		if(0)
+			button_icon_state = "car_count0"
```
Comment (AshCorr) [inline code/modules/vehicles/car/car_actions.dm]: God
Damn
You
Did
It
Again


button.icon_state = "car_count[min(count, 10)]`

kill this unholy switch.

File: code/modules/vehicles/vehicle.dm [hunk 3]
```diff
@@ -95,7 +95,7 @@
 	if(user.incapacitated())
 		unbuckle_mob(user)
 
-	if(keycheck(user))
+	if(keycheck(user) || istype(src, /obj/vehicle/car)) //Cars dont use this horrible shitty check
```
Comment (AshCorr) [inline code/modules/vehicles/vehicle.dm]: Cant you just set `keytype = false` in the type definition?

keycheck returns true if `keytype == false`

Comment (Qustinnus) [inline code/modules/vehicles/vehicle.dm]: no because cars use keys, just in a different way.

Comment (Qustinnus) [inline code/modules/vehicles/vehicle.dm]: (for starting the car)

File: code/modules/vehicles/car/car_actions.dm (hunk 2 above)
Comment (ghost) [reply code/modules/vehicles/car/car_actions.dm]: i told you its placeholder :^)))))))

**Comment by ghost** [comment]
@AshCorr AshCorr deleted a comment from Qustinnus an hour ago
 @AshCorr AshCorr deleted a comment from Qustinnus an hour ago

censorship, dont make me call goof to this PR he is very anti censor :^)

File: code/modules/vehicles/vehicle.dm (hunk 3 above)
Comment (Cruix) [inline code/modules/vehicles/vehicle.dm]: If cars do not use this check, override the keycheck() proc to always return true. If you want to reuse the keycheck proc somewhere else, make a wrapper proc that just calls keycheck() for vanilla vehicles and returns true for cars.

File: code/modules/vehicles/car/basic_car.dm [hunk 4]
```diff
@@ -0,0 +1,6 @@
+/obj/vehicle/car/basic
+	name = "Tourist car"
```
Comment (Cruix) [inline code/modules/vehicles/car/basic_car.dm]: This should be lower case because it is an improper noun

File: code/modules/vehicles/car/car.dm [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+	var/last_enginesound_time //To prevent sound spam
+	var/engine_sound_length = 20 //Set this to the length of the engine sound
+	var/engine_sound = 'sound/effects/carrev.ogg'
+
+	var/ramming = FALSE //Whether or not this car is ramming people.
+	var/last_crash_time //to prevent double-crashing into walls.
+	var/list/ramming_sounds = list() //Sounds for when you hit a person
+	var/list/crash_sounds = list()  //Sounds for when you crash into a structure
+
+	var/can_load_mobs = FALSE //Whether or not this car can have people in it's trunk, for meme vehicles
```
Comment (Cruix) [inline code/modules/vehicles/car/car.dm]: it's should be its

File: code/modules/vehicles/car/car_actions.dm [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+		playsound(car.loc, 'sound/effects/car_start.ogg', 50)
+		car.on = TRUE
+	UpdateButtonIcon()
+
+/datum/action/innate/car/dump_load
+	name = "Dump contents"
+	button_icon_state = "car_dump"
+
+/datum/action/innate/car/dump_load/Activate()
+	car.visible_message("<span class='danger'>[car] dumps all of it's contents on the floor.	</span>")
```
Comment (Cruix) [inline code/modules/vehicles/car/car_actions.dm]: it's should be its

File: code/modules/vehicles/car/car.dm [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+	if(driver_visible) //Buckling
+		driver.unbuckle_mob(driver, forced)
+	else //Using contents
+		driver.forceMove(newloc)
+	driver = null
+
+/obj/vehicle/car/proc/load_mob(mob/M)
+	if(!istype(M))
+		return
+	if(M && M in range(1, src))
```
Comment (Cruix) [inline code/modules/vehicles/car/car.dm]: (M.z == z) && (get_dist(M, src) <= 1 ) is faster

File: code/modules/vehicles/car/car.dm [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+			user.visible_message("<span class='danger'>[user] stuffs [target] into [src]</span>")
+			load_mob(target)
+
+/obj/vehicle/car/MouseDrop(atom/over_object)
+	if(driver)
+		usr.visible_message("<span class='danger'>[usr] starts dragging [driver] out of [src]</span>")
+		if(do_after(usr, 20, target = src))
+			usr.visible_message("<span class='danger'>[usr] drags [driver] out of [src]</span>")
+			exit_car()
+			dump_contents()
```
Comment (Cruix) [inline code/modules/vehicles/car/car.dm:104]: You can drag yourself out of the car, I do not know if this is intended.

File: code/modules/vehicles/car/car.dm [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+	user.visible_message("<span class='danger'>[user] hits [name]. Nothing happens</span>","<span class='danger'>You hit [name] with no visible effect.</span>")
+	return
+
+/obj/vehicle/car/narsie_act()
+	take_damage(120)
+
+/obj/vehicle/car/ratvar_act()
+	take_damage(120)
+
+/obj/vehicle/car/hitby(atom/movable/A as mob|obj) //wrapper
```
Comment (Cruix) [inline code/modules/vehicles/car/car.dm]: the as mob|obj does not do anything in this context, everything else that does this proc this way are old and wrong.

File: code/modules/vehicles/car/car.dm [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+	if(amount)
+		health -= amount
+		update_health()
+
+/obj/vehicle/car/proc/update_health()
+	if(health < 0)
+		qdel(src)
+
+/obj/vehicle/car/proc/enter_car(mob/living/carbon/human/H)
+	if(H && H.client && H in range(1))
```
Comment (Cruix) [inline code/modules/vehicles/car/car.dm]: Same as the other range(1) check

Comment (Cruix) [reply code/modules/vehicles/car/car.dm]: This proc should also have an ishuman() check

File: code/modules/vehicles/car/car.dm [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+		return
+	RemoveActions(driver)
+	if(driver_visible) //Buckling
+		driver.unbuckle_mob(driver, forced)
+	else //Using contents
+		driver.forceMove(newloc)
+	driver = null
+
+/obj/vehicle/car/proc/load_mob(mob/M)
+	if(!istype(M))
```
Comment (Cruix) [inline code/modules/vehicles/car/car.dm]: These should be checking for mob/living, so you cannot shove ghosts into the car if you can somehow see them

**Review by Cruix — CHANGES_REQUESTED** [review]
10/10 crashed into a vending machine and got beaten to death by the mime.

File: code/modules/vehicles/car/car.dm (hunk 8 above)
Comment (ghost) [reply code/modules/vehicles/car/car.dm:104]: 🤔 good point

**Comment by ghost** [comment]
i'll finish my 3 PRs next week, going out drinking all week this week so I won't have time to fix the requested changes for now :^)

**Comment by KingPhilipIII** [comment]
The one time I used a goon clown car I tried to abduct the clown and he spaced me when he got out.

**Comment by ghost** [comment]
Ready for review

**Comment by Jc1469** [comment]
I'm fine with this being added, but I have to question the... blatancy of it.

**Comment by ghost** [comment]
what blatancy

**Comment by Kmc2000** [comment]
Yeah the clown car is a massive callout "HEY IM A SYNDIE LOOK AT ME" but it's not like it's black and red or anything, so pubbies will have no idea where the fuck it came from and that gets a 👍  from me and answering their tickets

File: code/modules/uplink/uplink_item.dm [hunk 12]
```diff
@@ -1064,14 +1064,21 @@ var/list/uplink_items = list() // Global list so we only initialize this once.
...
 	//restricted_roles = list("Assistant")
 
+/datum/uplink_item/role_restricted/clowncar
+	name = "Clown Car"
+	category = "item category"
+	desc = "The finest vehicle in the Clown Planet selection. This baby can fit atleast 4932004 unfortunate victims and comes with a built in horn. Drive around in style \
+	as you run down innocent mobs of people as they run from this beautiful creation of The Honkmother."
+	item = /obj/vehicle/car/clown
+	cost = 12
+	list/restricted_roles = list("Clown")
```
Comment (Cruix) [inline code/modules/uplink/uplink_item.dm]: Remove the list/
Right now even non-clowns can buy this

File: code/modules/uplink/uplink_item.dm [hunk 13]
```diff
@@ -1064,14 +1064,21 @@ var/list/uplink_items = list() // Global list so we only initialize this once.
...
-
 // Role-specific items
 /datum/uplink_item/role_restricted
 	category = "Role-Restricted"
 	exclude_modes = list(/datum/game_mode/nuclear)
 	//restricted_roles = list("Assistant")
 
+/datum/uplink_item/role_restricted/clowncar
+	name = "Clown Car"
+	category = "item category"
```
Comment (Cruix) [inline code/modules/uplink/uplink_item.dm]: Remove this line

File: code/modules/vehicles/vehicle.dm (hunk 3 above)
Comment (Cruix) [reply code/modules/vehicles/vehicle.dm]: Again, do not use istype(src)

**Comment by Cruix** [comment]
Make sure when you fix it that only clowns can buy it

File: code/modules/uplink/uplink_item.dm (hunk 13 above)
Comment (ghost) [reply code/modules/uplink/uplink_item.dm]: sorry im acoustic

File: code/modules/uplink/uplink_item.dm (hunk 12 above)
Comment (ghost) [reply code/modules/uplink/uplink_item.dm]: sorry im acoustic

**Comment by ghost** [comment]
Fixed and tested @Cruix

File: code/modules/vehicles/vehicle.dm (hunk 3 above)
Comment (Cruix) [reply code/modules/vehicles/vehicle.dm]: Still an issue

**Comment by ghost** [comment]
we need more head coders

**Comment by Kmc2000** [comment]
Apply then

**Comment by ghost** [comment]
>implying id get accepted

**Comment by Cruix** [comment]
Remove the istype(src)

**Comment by ghost** [comment]
i am doing that rn

**Comment by ghost** [comment]
but besides that is there anything major i dont like continiously doing small fixes and wait for reviews

**Comment by Cruix** [comment]
I am not going to be able to test it right now so I cannot say.
Also, life hack: If you find all the bugs in your PRs yourself, and address all points of the first review immediately, you only have to wait for one review.

File: code/modules/vehicles/vehicle.dm (hunk 3 above)
Comment (ghost) [reply code/modules/vehicles/vehicle.dm]: alright I'm tired af now I'll fix this tommorow, drop any extra things i need to do in here too i hate doing fixes and then waiting, I prefer sitting down to fix 50 things instead

**Comment by ghost** [comment]
hahahahahahahaha that would imply I'd have to test.

but in all seriousness, the only bug that's really my fault for not getting is the dragging yourself out of the car one :^) the rest I legit had no good way of finding out about.

**Comment by Cruix** [comment]
You could have actually tried to buy one with the traitor uplink.

File: code/modules/vehicles/vehicle.dm (hunk 3 above)
Comment (Cruix) [reply code/modules/vehicles/vehicle.dm]: If I knew about anything else I would have commented on it already.

**Comment by ghost** [comment]
I did, I forgot I re-did the uplink stuff later though and fucked up in the progress.

**Comment by ghost** [comment]
"I'm going to fix the remaining bugs tommorow"

"time to testmerge"

good shit. especially the part where i wasn't informed

**Comment by Cruix** [comment]
If you ask me for a review I expect that you have already tested it as thoroughly as you can, and that if I do not find anything wrong with it that it is ready to be merged. The only issue remaining is a code quality concern, not an in game bug. If you do not want me to merge something, all you have to do is say so.

There have never been scheduled times for server updates or testmerges, and the testmerge label is more than enough notification for anyone who watches github. My going through the effort of putting your code on the server early for player feedback is in no way a method of persecution.

**Comment by ghost** [comment]
I changed the code yesterday, as I was working on a fix already but went to bed before finishing it. This means that there could have been changes that could've made the car completely dysfunctional, luckily the car still worked if you were holding the bikehorn, but it's still horribly buggy and shitty, and not the way its ment to be.

**Comment by ghost** [comment]
feel free to finish this or wait for the TG vehicle refactor and port that instead.

**Pull request closed without merging**